apisx/v1alpha1/xmesh_types.go

```diff
@@ -0,0 +1,183 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	gatewayapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=gateway-api,scope=Cluster,shortName=mesh
+// +kubebuilder:subresource:status
+// +kubebuilder:storageversion
+// +kubebuilder:printcolumn:name="Accepted",type=string,JSONPath=`.status.conditions[?(@.type=="Accepted")].status`
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// XMesh defines mesh-wide characteristics of a GAMMA-compliant service mesh.
+type XMesh struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of XMesh.
+	// +required
+	Spec MeshSpec `json:"spec"`
+
+	// Status defines the current state of XMesh.
+	//
+	// <gateway:util:excludeFromCRD>
+	// Implementations MUST populate status on all Mesh resources which
+	// specify their controller name.
+	// </gateway:util:excludeFromCRD>
+	//
+	// +kubebuilder:default={conditions: {{type: "Accepted", status: "Unknown", reason:"Pending", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}}
+	// +optional
+	Status MeshStatus `json:"status,omitempty"`
+}
+
+// MeshSpec defines the desired state of an XMesh.
+type MeshSpec struct {
+	// ControllerName is the name of a controller that is managing Gateway API
+	// resources for mesh traffic management. The value of this field MUST be a
+	// domain prefixed path.
+	//
+	// Example: "example.com/awesome-mesh".
+	//
+	// This field is not mutable and cannot be empty.
+	//
+	// Support: Core
+	//
+	// +kubebuilder:validation:XValidation:message="Value is immutable",rule="self == oldSelf"
+	// +required
+	ControllerName gatewayapiv1.GatewayController `json:"controllerName"`
+
+	// ParametersRef is an optional reference to a resource that contains
+	// implementation-specific configuration for this Mesh. If no
+	// implementation-specific parameters are needed, this field MUST be
+	// omitted.
+	//
+	// ParametersRef can reference a standard Kubernetes resource, i.e.
+	// ConfigMap, or an implementation-specific custom resource. The resource
+	// can be cluster-scoped or namespace-scoped.
+	//
+	// If the referent cannot be found, refers to an unsupported kind, or when
+	// the data within that resource is malformed, the Mesh MUST be rejected
+	// with the "Accepted" status condition set to "False" and an
+	// "InvalidParameters" reason.
+	//
+	// Support: Implementation-specific
+	//
+	// +optional
+	ParametersRef *gatewayapiv1.ParametersReference `json:"parametersRef,omitempty"`
```

Comment:
I vaguely remember Rob saying omitempty is not longer required/needed for optional fields?

@robscott ?

Reply:
Not sure -- I'm going to leave `omitempty` in until we see the rest of ecosystem dropping it though!

Reply:
It's actually the opposite - we use `omitempty` everywhere now, including required fields.

Reply:
🫣